py/htm/algorithms/anomaly_likelihood.py

```diff
@@ -15,205 +16,81 @@
-      intervals.
-
-    :param reestimationPeriod: (int) how often we re-estimate the Gaussian
-      distribution. The ideal is to re-estimate every iteration but this is a
-      performance hit. In general the system is not very sensitive to this
-      number as long as it is small relative to the total number of records
-      processed.
+    Argument period is the number of samples to use to calculate the
+    distribution of anomaly scores.
     """
-    if historicWindowSize < estimationSamples:
-      raise ValueError("estimationSamples exceeds historicWindowSize")
-
-    self._iteration = 0
-    self._historicalScores = collections.deque(maxlen=historicWindowSize)
-    self._distribution = None
-
-
-    if claLearningPeriod != None:
-      print("claLearningPeriod is deprecated, use learningPeriod instead.")
-      self._learningPeriod = claLearningPeriod
-    else:
-      self._learningPeriod = learningPeriod
-
-    self._probationaryPeriod = self._learningPeriod + estimationSamples
-    self._reestimationPeriod = reestimationPeriod
-
-
-  def __eq__(self, o):
-    # pylint: disable=W0212
-    return (isinstance(o, AnomalyLikelihood) and
-            self._iteration == o._iteration and
-            self._historicalScores == o._historicalScores and
-            self._distribution == o._distribution and
-            self._probationaryPeriod == o._probationaryPeriod and
-            self._learningPeriod == o._learningPeriod and
-            self._reestimationPeriod == o._reestimationPeriod)
-    # pylint: enable=W0212
-
-
-  def __str__(self):
-    return ("AnomalyLikelihood: %s %s %s %s %s %s" % (
-      self._iteration,
-      self._historicalScores,
-      self._distribution,
-      self._probationaryPeriod,
-      self._learningPeriod,
-      self._reestimationPeriod) )
+    self.period   = float(period)
+    self.alpha    = 1.0 - math.exp(-1.0 / self.period)
+    self.mean     = 1.0
+    self.var      = 0.0003
+    self.std      = math.sqrt(self.var)
+    self.prev     = 0.0
+
+  def compute(self, anomaly):
+    """ Returns the probability that an anomaly has occurred.
+
+    Calculates the log-likelihood of the probability that the given anomaly
+    score is greater than a historical distribution of anomaly scores. Then
+    updates the historical distribution to include the given anomaly score.
+    """
+    likelihood = self.get_likelihood(anomaly)
+    self.add_record(anomaly)
+    combined = 1 - (1 - likelihood) * (1 - self.prev)
+    self.prev = likelihood
+    return self.get_log_likelihood(combined)
 
+  def add_record(self, anomaly):
+    """
+    """
+    # http://web.archive.org/web/http://people.ds.cam.ac.uk/fanf2/hermes/doc/antiforgery/stats.pdf
+    # Skip to Chapter 9.
+    diff      = anomaly - self.mean
+    incr      = self.alpha * diff
+    self.mean = self.mean + incr
+    self.var  = (1.0 - self.alpha) * (self.var + diff * incr)
+    if True:
```

Comment:
?

Reply:
Yes, the code in the `if True:` block does a different thing than the rest of that method does so I wanted it in its own block of code, for readability purposes.

Also, you *could* replace True with False and see what happens when you don't have a minimum variance.  I played with the minimum-variance a little bit when I was developing this, and came to the conclusion that it should be always on.